File: go-backend-interview-prep/08_http_and_backend/09_rate_limit_basics/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"sync"
	"time"
)

// TokenBucket implements a simple rate limiter.
type TokenBucket struct {
	mu       sync.Mutex
	tokens   int
	max      int
	interval time.Duration
}

func NewTokenBucket(max int, refillInterval time.Duration) *TokenBucket {
	tb := &TokenBucket{
		tokens:   max,
		max:      max,
		interval: refillInterval,
	}
	// Refill tokens periodically
	go func() {
		ticker := time.NewTicker(refillInterval)
		defer ticker.Stop()
		for range ticker.C {
			tb.mu.Lock()
			tb.tokens = tb.max
			tb.mu.Unlock()
		}
	}()
	return tb
}

func (tb *TokenBucket) Allow() bool {
	tb.mu.Lock()
	defer tb.mu.Unlock()
	if tb.tokens > 0 {
		tb.tokens--
		return true
	}
	return false
}

func (tb *TokenBucket) Remaining() int {
	tb.mu.Lock()
	defer tb.mu.Unlock()
	return tb.tokens
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

// rateLimitMiddleware wraps a handler with token bucket checking.
func rateLimitMiddleware(limiter *TokenBucket, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !limiter.Allow() {
			w.Header().Set("Retry-After", fmt.Sprintf("%.0f", limiter.interval.Seconds()))
			writeJSON(w, http.StatusTooManyRequests, map[string]string{
				"error":       "rate limit exceeded",
				"retry_after": fmt.Sprintf("%.0fs", limiter.interval.Seconds()),
			})
			return
		}
		w.Header().Set("X-RateLimit-Remaining", fmt.Sprintf("%d", limiter.Remaining()))
		next.ServeHTTP(w, r)
	})
}

func apiHandler(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{
		"message": "API response",
	})
}

func main() {
	// 5 requests per 3 seconds
	limiter := NewTokenBucket(5, 3*time.Second)

	mux := http.NewServeMux()
	mux.HandleFunc("/api", apiHandler)

	handler := rateLimitMiddleware(limiter, mux)

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("Server on http://%s\n", addr)
	fmt.Println("  GET /api -> rate limited (5 req / 3s)")

	server := &http.Server{Handler: handler}

	// Self-test: send 8 requests rapidly
	go func() {
		time.Sleep(100 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo: 8 rapid requests ---")

		for i := 1; i <= 8; i++ {
			resp, err := http.Get(base + "/api")
			if err != nil {
				fmt.Printf("  request %d: error: %v\n", i, err)
				continue
			}
			var body map[string]string
			json.NewDecoder(resp.Body).Decode(&body)
			resp.Body.Close()
			remaining := resp.Header.Get("X-RateLimit-Remaining")
			fmt.Printf("  request %d: status=%d remaining=%s msg=%s\n",
				i, resp.StatusCode, remaining, body["message"]+body["error"])
		}

		// Wait for refill
		fmt.Println("\n  waiting 3s for token refill...")
		time.Sleep(3 * time.Second)

		resp, _ := http.Get(base + "/api")
		if resp != nil {
			resp.Body.Close()
			fmt.Printf("  after refill: status=%d\n", resp.StatusCode)
		}

		fmt.Println("\nDemo done.")
		server.Close()
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}

File: go-backend-interview-prep/08_http_and_backend/09_rate_limit_basics/main.py
"""Rate limit basics -- Python equivalent of the Go example.

Token bucket rate limiter with threading.Lock for concurrency safety.
"""

import json
import threading
import time
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler


class TokenBucket:
    def __init__(self, max_tokens: int, refill_interval: float):
        self.max_tokens = max_tokens
        self.tokens = max_tokens
        self.refill_interval = refill_interval
        self.lock = threading.Lock()
        # Refill tokens periodically
        self._start_refill()

    def _start_refill(self):
        def refill():
            while True:
                time.sleep(self.refill_interval)
                with self.lock:
                    self.tokens = self.max_tokens
        t = threading.Thread(target=refill, daemon=True)
        t.start()

    def allow(self) -> bool:
        with self.lock:
            if self.tokens > 0:
                self.tokens -= 1
                return True
            return False

    def remaining(self) -> int:
        with self.lock:
            return self.tokens


# Global limiter: 5 requests per 3 seconds
limiter = TokenBucket(max_tokens=5, refill_interval=3.0)


def json_response(handler, status: int, data: dict, extra_headers=None) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    if extra_headers:
        for k, v in extra_headers.items():
            handler.send_header(k, v)
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/api":
            if not limiter.allow():
                json_response(self, 429, {
                    "error": "rate limit exceeded",
                    "retry_after": f"{limiter.refill_interval:.0f}s",
                }, extra_headers={
                    "Retry-After": str(int(limiter.refill_interval)),
                })
                return

            json_response(self, 200, {"message": "API response"}, extra_headers={
                "X-RateLimit-Remaining": str(limiter.remaining()),
            })
            return

        json_response(self, 404, {"error": "not found"})

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), Handler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Server on {base}")
    print("  GET /api -> rate limited (5 req / 3s)")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    print("\n--- Demo: 8 rapid requests ---")
    for i in range(1, 9):
        try:
            resp = urllib.request.urlopen(f"{base}/api", timeout=5)
            data = json.loads(resp.read())
            remaining = resp.headers.get("X-RateLimit-Remaining", "?")
            print(f"  request {i}: status={resp.status} remaining={remaining} msg={data.get('message', '')}")
        except urllib.error.HTTPError as e:
            data = json.loads(e.read())
            print(f"  request {i}: status={e.code} msg={data.get('error', '')}")

    print(f"\n  waiting 3s for token refill...")
    time.sleep(3)

    resp = urllib.request.urlopen(f"{base}/api", timeout=5)
    print(f"  after refill: status={resp.status}")

    server.shutdown()
    print("\nDone.")


if __name__ == "__main__":
    main()
